topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java

```diff
@@ -616,6 +616,63 @@ public void handle(Future<Void> fut) {
         inFlight.enqueue(topicName, resultHandler, futureHandler);
     }
 
+    void onTopicPartitionsChanged(TopicName topicName, Handler<AsyncResult<Void>> resultHandler) {
+        Handler<Future<Void>> futureHandler = new Reconciliation("onTopicPartitionsChanged") {
+            @Override
+            public void handle(Future<Void> fut) {
+
+                Handler<AsyncResult<TopicMetadata>> handler = new Handler<AsyncResult<TopicMetadata>>() {
+
+                    private final BackOff backOff = new BackOff();
+
+                    @Override
+                    public void handle(AsyncResult<TopicMetadata> metadataResult) {
+
+                        if (metadataResult.succeeded()) {
+
+                            // getting topic metadata from Kafka
+                            Topic kafkaTopic = TopicSerialization.fromTopicMetadata(metadataResult.result());
+
+                            // getting topic information from the private store
+                            topicStore.read(topicName, topicResult -> {
+
+                                // if partitions aren't changed on Kafka yet, we retry with exponential backoff
+                                if (topicResult.result().getNumPartitions() == kafkaTopic.getNumPartitions()) {
+
+                                    long delay;
```

Comment:
There's a lot of shared code here with the topic creation case -- we can probably factor out a common abstract class to encapsulate the timing

Reply:
Agree on that but are you thinking about abstracting the `Controller` or having something like a `HandlerWithBackoff<E>` class (derived from the `Handler<E>`) for supporting such a feature. (don't consider the name, it's just for giving you an idea :-))

Reply:
I was thinking more along the lines of a `HandlerWithBackOff`. I think in both cases we're retrying getting the topic metadata, so a `TopicMetadataHandler`, with an abstract method for determining when the condition it met (and no more retries are needed) and another one for how to handle the `MaxAttemptsExceededException` should work.

Reply:
Yes exactly, it was what I had in mind. Working on that ;)

Reply:
@tombentley what do you think after the latest commit ? Btw I have an idea about having a more generic handler with retry mechanism with backoff but maybe it's worth a different PR (in the future).